libdnf/rpm/solv/goal_private.cpp

```diff
@@ -578,5 +578,41 @@ void GoalPrivate::reset_protected_packages() {
     protected_packages.reset();
 }
 
+transaction::TransactionItemReason GoalPrivate::get_reason(Id id) {
+    //solver_get_recommendations
+    if (!libsolv_solver)
+        throw UnresolvedGoal();
+    Id info;
+    int reason = solver_describe_decision(libsolv_solver, id, &info);
+
+    if ((reason == SOLVER_REASON_UNIT_RULE ||
+         reason == SOLVER_REASON_RESOLVE_JOB) &&
+        (solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_JOB ||
+         solver_ruleclass(libsolv_solver, info) == SOLVER_RULE_BEST))
+        return transaction::TransactionItemReason::USER;
+    if (reason == SOLVER_REASON_CLEANDEPS_ERASE)
+        return transaction::TransactionItemReason::CLEAN;
```

Comment:
The `CLEAN` value is marked as "hawkey compatibility", it would seem we want to get rid of it (at least I don't see its purpose)?

Reply:
I use the reason clean because it is needed. For remove action we need 3 reasons:
TransactionItemReason::USER
TransactionItemReason::DEPENDENCY (for removal of dependent packages)
TransactionItemReason::CLEAN (for removal of unused dependenies)

The marker about compatibility is incorrect and I will remove it.

Reply:
What's the difference between DEPENDENCY and CLEAN? If something depends on a package you can't remove it. If nothing depends on a package, you can remove it and it's in effect the clean operation?

Reply:
Installed packages `A`, `B`, `C`
`A` requires `B`
`B` requires `C`
`dnf remove B` => remove `A`, `B`, `C`
`B` as USER
`C` as CLEAN
`A` as DEP

Reply:
Oh... So when you install A (reason USER), B is installed with reason DEP.

When you remove B, it is removed as USER and A is removed as DEP? That is just confusing.

Reply:
Original name was not DEPENDENCY but DEP. DEP means dependency (forward action) or dependent (reverse action - remove).

Reply:
That's just the same or worse, using the same value for two different meanings and hiding it behind and ambiguous name abbreviation (which could be technically labeled as "corect") :nerd_face:

Reply:
Anyway the discussion started about CLEAN and that part is resolved.